### PR TITLE
Handle topic renaming when checking destination partitions

### PR DESCRIPTION
**Pull request by parafiend**
Add support for RegexRouter transforms and topic name prefixes and
suffixes when checking the destination cluster for topic name.
Previously the KafkaMonitor only checked for the original topic name,
even if messages would be replicated to a different topic.

Prefixes and suffixes are applied first, since transforms are only
applied to records after being returned from the MirusSourceTask::poll
method.

Other routers aren't currently supported. The Kafka Connect
TimestampRouter routes based of information within each message, so
the required topic name can't be determined ahead of time. The
KafkaMonitor assumes any transform that contains the phrase "Router"
would rename topics and logs a warning about mirroring when identified.

### CODE REVIEW

File: src/test/java/com/salesforce/mirus/KafkaMonitorTest.java [hunk 1]
```diff
@@ -138,4 +140,115 @@ public void shouldNotReBalanceIfOnlySourcePartitionOrderHasChanged() {
             new PartitionInfo("topic1", 0, null, null, null)));
     assertThat(kafkaMonitor.partitionsChanged(), is(false));
   }
+
+  @Test
+  public void shouldApplyTopicRenameTransforms() {
+    Map<String, String> properties = new HashMap<>();
```
Comment (parafiend) [inline src/test/java/com/salesforce/mirus/KafkaMonitorTest.java]: There's a fair bit of duplication in the setup for these tests, recommendations on reducing that appreciated.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 2]
```diff
@@ -89,6 +96,51 @@
...
+
+  private List<Transformation<SourceRecord>> validateTransformations(
+      List<Transformation<SourceRecord>> transformations) {
+    List<Transformation<SourceRecord>> regexRouters = new ArrayList<>();
+    for (Transformation<SourceRecord> transform : transformations) {
+      String transformName = transform.getClass().getName();
+      if (transform instanceof RegexRouter) {
+        regexRouters.add(transform);
+        // Slightly awkward check to see if any other routing transforms are configured
+      } else if (transform.getClass().getName().contains("Router")) {
```
Comment (adammelliott) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Do you mean to be checking the full name of the class including package, or are you more interested in the class' simple name? In any case, you already obtained the name once already as `transformName`.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 3]
```diff
@@ -217,6 +269,8 @@ boolean partitionsChanged() {
     Map<Boolean, List<TopicPartition>> partitionedSourceIds =
         sourcePartitionList
             .stream()
+            .map(this::applyTopicNameConfigsToTopic)
+            .map(this::applyRoutersToTopic)
```
Comment (adammelliott) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: I kind of feel that prefix and suffix should be added _after_ the transformation router has executed, not before, but I'm not sure that's anything more than a personal opinion--and in any case, it should match what is already happening in the other part of the code. @pdavidson100 what do you think?

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 4]
```diff
@@ -89,6 +96,51 @@
...
+    List<Transformation<SourceRecord>> regexRouters = new ArrayList<>();
+    for (Transformation<SourceRecord> transform : transformations) {
+      String transformName = transform.getClass().getName();
+      if (transform instanceof RegexRouter) {
+        regexRouters.add(transform);
+        // Slightly awkward check to see if any other routing transforms are configured
+      } else if (transform.getClass().getName().contains("Router")) {
+        logger.error(
+            "Unsupported Router Transform %s found. Destination topic checking may not work.",
+            transformName);
```
Comment (adammelliott) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Should there be a way to disable destination topic checking in this case?

File: src/main/java/com/salesforce/mirus/config/SourceConfigDefinition.java [hunk 5]
```diff
@@ -111,6 +112,13 @@ public static ConfigDef configDef() {
     for (SourceConfigDefinition f : SourceConfigDefinition.values()) {
       configDef = configDef.define(f.key, f.type, f.defaultValue, f.importance, f.doc);
     }
+
+    // Share name and all transform config definitions from ConnectorConfig
+    for (ConfigKey key : ConnectorConfig.configDef().configKeys().values()) {
+      if ("Transforms".equals(key.group) || ConnectorConfig.NAME_CONFIG.equals(key.name)) {
```
Comment (adammelliott) [inline src/main/java/com/salesforce/mirus/config/SourceConfigDefinition.java]: Is this really case sensitive?

File: src/test/java/com/salesforce/mirus/KafkaMonitorTest.java [hunk 6]
```diff
@@ -138,4 +140,115 @@ public void shouldNotReBalanceIfOnlySourcePartitionOrderHasChanged() {
...
+    monitor.partitionsChanged();
+    List<Map<String, String>> result = monitor.taskConfigs(3);
+
+    List<List<TopicPartition>> partitions =
+        result
+            .stream()
+            .map(i -> new TaskConfig(i).getInternalTaskPartitions())
+            .map(TopicPartitionSerDe::parseTopicPartitionList)
+            .collect(Collectors.toList());
+    assertThat(partitions.size(), is(1));
```
Comment (adammelliott) [inline src/test/java/com/salesforce/mirus/KafkaMonitorTest.java]: If you're only doing a size-check on the collection, what is the value of running it through the transformation?

File: src/test/java/com/salesforce/mirus/config/SourceConfigTest.java [hunk 7]
```diff
@@ -34,6 +41,58 @@ public void setUp() {
...
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof SimpleTransformation) {
+        SimpleTransformation other = (SimpleTransformation) obj;
+        if (this.magicNumber == other.magicNumber) {
+          return true;
+        }
+      }
+      return false;
```
Comment (adammelliott) [inline src/test/java/com/salesforce/mirus/config/SourceConfigTest.java]: Given that this is a single-use test class, can you just condense `hashCode` and `equals` into one- or two-liners for the sake of reducing visual noise?

File: src/test/java/com/salesforce/mirus/config/SourceConfigTest.java [hunk 8]
```diff
@@ -9,20 +9,27 @@
...
+import java.util.List;
 import java.util.Map;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.transforms.Transformation;
 import org.junit.Before;
 import org.junit.Test;
 
-public class SourceConfigTest {
+public class SourceConfigTest<R extends ConnectRecord<R>> {
```
Comment (adammelliott) [inline src/test/java/com/salesforce/mirus/config/SourceConfigTest.java]: The type parameter on the test is a little weird. Is there a better way?

File: src/test/java/com/salesforce/mirus/KafkaMonitorTest.java (hunk 1 above)
Comment (adammelliott) [reply src/test/java/com/salesforce/mirus/KafkaMonitorTest.java]: It looks like you could create two helper methods to contribute prefix/suffix config and regex rename config respectively to the properties instance. You'd have to use the values from the test that contains both; but those, I think, would work independently. Then just call `addPrefixSuffixConfig` or `addRegexTransformConfig` as needed in your test cases.

I think the rest of the stuff is hard to refactor away in a way that wouldn't obscure the workings of the test.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java (hunk 3 above)
Comment (parafiend) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: I agree that's the more logical way, but [Kafka Connect applies transformations after the SourceTask finishes processing](https://github.com/apache/kafka/blob/af634a4a98eaa2457752e3f2841720020e0e9ad0/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L218) each batch of records. Transformations are applied inside of `WorkerSourceTask::sendRecords()`, which is only called after `MirusSourceTask::poll()` is called from `WorkerSourceTask::poll()`, so the Mirus Connector and Task aren't able to affect records after they're transformed.

#31 suggests deprecating the prefix/suffix options entirely since RegexRouter supports that fairly easily, which I think the best long-term plan.

File: src/main/java/com/salesforce/mirus/config/SourceConfigDefinition.java (hunk 5 above)
Comment (parafiend) [reply src/main/java/com/salesforce/mirus/config/SourceConfigDefinition.java]: The source is a protected constant inside Kafka Connect's [ConnectorConfig](https://github.com/apache/kafka/blob/a2e87feb8b1db8200ca3a34aa72b0802e8f61096/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java#L58), which is indeed capitalized. I could switch to `equalsIgnoreCase()` if we're worried about capitalization changing

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 9]
```diff
@@ -89,6 +96,51 @@
...
+  }
+
+  private List<Transformation<SourceRecord>> validateTransformations(
+      List<Transformation<SourceRecord>> transformations) {
+    List<Transformation<SourceRecord>> regexRouters = new ArrayList<>();
+    for (Transformation<SourceRecord> transform : transformations) {
+      String transformName = transform.getClass().getName();
+      if (transform instanceof RegexRouter) {
+        regexRouters.add(transform);
+        // Slightly awkward check to see if any other routing transforms are configured
```
Comment (pdavidson100) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: This make me unhappy, but nothing better springs to mind. We will need to be clear about the behaviour in the documentation.  We should update the JavaDoc for KafkaMonitor to explain, and maybe a note in the Readme.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java (hunk 3 above)
Comment (pdavidson100) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Because this is a Source Connector the transformations are applied _after_ we return the records to the framework, so necessarily after the topic name transformation happens in the connector code.  I think @parafiend has the order right.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 10]
```diff
@@ -89,6 +96,51 @@
...
+      } else if (transform.getClass().getName().contains("Router")) {
+        logger.error(
```
Comment (pdavidson100) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: I think as long as we document this behaviour clearly it's not really an error. I think we should make it a `warn`, and say something like "Destination topic checking is not supported with router transform %s enabled. Please ensure all possible destination topics exist".  Also, probably should disable all destination topic checking in that case as might end up checking for topics that are not intended to exist?

File: src/test/java/com/salesforce/mirus/KafkaMonitorTest.java (hunk 6 above)
Comment (parafiend) [reply src/test/java/com/salesforce/mirus/KafkaMonitorTest.java]: If the transformation isn't properly applied, after `monitor.partitionsChanged()` is called above, no partitions will be mirrored and the collection will be empty. Based off the existing test checking that missing partitions aren't mirroed.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java (hunk 4 above)
Comment (pdavidson100) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Yeah, I think that's what I'm getting at with my earlier comment.

File: src/test/java/com/salesforce/mirus/config/SourceConfigTest.java (hunk 8 above)
Comment (pdavidson100) [reply src/test/java/com/salesforce/mirus/config/SourceConfigTest.java]: Why not just make `SimpleTransformation` a simple non-nested class? Can probably still be in the same file if you make it package-private instead of public.

File: src/test/java/com/salesforce/mirus/config/SourceConfigTest.java [hunk 11]
```diff
@@ -52,4 +111,19 @@ public void destinationBootstrapShouldBeAvailable() {
...
+  public void transformationsShouldBeAvailable() {
+    Map<String, String> properties = new HashMap<>();
+    properties.put("name", "connector");
+    properties.put("transforms", "a");
+    properties.put("transforms.a.type", SimpleTransformation.class.getName());
+    properties.put("transforms.a.magic.number", "45");
+    SourceConfig configWithTransform = new SourceConfig(properties);
+
+    List<Transformation<R>> transformations = configWithTransform.transformations();
+    SimpleTransformation<R> transform = new SimpleTransformation<R>();
```
Comment (pdavidson100) [inline src/test/java/com/salesforce/mirus/config/SourceConfigTest.java]: Trivia: the second <R> can be <>

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java (hunk 4 above)
Comment (parafiend) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: That could be useful. Do you think it's worth limiting it to just this case ("Topics might be rerouted and we're not sure where, so always try to push them") or just having a global switch to disable destination checking? @pdavidson100 ?

File: src/test/java/com/salesforce/mirus/KafkaMonitorTest.java (hunk 6 above)
Comment (pdavidson100) [reply src/test/java/com/salesforce/mirus/KafkaMonitorTest.java]: I guess you're relying on the destination partition being pruned out out of the result set if the expected names don't match. Would be nice if there was a way to express that more directly in the test - is hard to follow what's going on here. 

Not sure, but would it help to externalize the `MissingPartitionsJmxReporter`class? That way we we can directly test whether partitions were considered missing?

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java (hunk 4 above)
Comment (pdavidson100) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: I think it would be safer to automatically disable in this case in combination with a clear warning, something like "WARN Destination topic validation will be disabled while unsupported router transform "%s" is enabled. Please ensure all possible destination topics exist."

Comment (pdavidson100) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Reconsidering,  might be even safer if we made destination checking configurable per connector and then make the connector hard fail in this situation with the message "Unsupported configuration, please disable destination checking in connector configuration".  That way we fail fast.  Thoughts?

Comment (parafiend) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: I like the fail fast best, although it may break some existing configurations. Technically Mirus is still pre-1.0, do we have a versioning/breaking changes plan?

Comment (pdavidson100) [reply src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Since we're still pre-1.0 I think we are OK to introduce this change, especially since we would likely just be surfacing an existing issue (and the fix is a simple config update called out in the error message). 

We absolutely need to get serious about issuing release notes covering important differences like this. I think we should add some release notes for 0.2 (the next feature release) to  this PR.

**Comment by parafiend** [comment]
@adammelliott @pdavidson100 I'll open a separate PR for the deprecation/removal of prefix/suffix configs

**Comment by parafiend** [comment]
@pdavidson100 @adammelliott Take another look. I've saved the following as draft release notes in Github:

> * Add support for `RegexRouter` to destination topic checking
> * Add `disable.destination.topic.checking` configuration option to disable checking for destination topic existence. This also allows for the unchecked usage of other topic re-routing transformations.

File: README.md [hunk 12]
```diff
@@ -200,6 +200,17 @@ Worker Properties file provided at startup. The Kafka Producer instances created
 can also be configured using a `producer.` prefix on the standard
 [Kafka Producer properties](https://kafka.apache.org/documentation/#producerconfigs).
 
+## Destination Topic Checking
+
+By default, Mirus checks that the destination topic exists in the destination Kafka cluster before
+stating to replicate data to it. This feature can be disabled by setting the
+[disable.destination.topic.checking](src/main/java/com/salesforce/mirus/config/SourceConfigDefinition.java#L66)
+config option to `true`.
```
Comment (adammelliott) [inline README.md]: I think I prefer "positive" rather than "negative" feature flags: so `enable...`, defaulting to `true`.

Comment (adammelliott) [reply README.md]: Same with the boolean flags within the code.

File: src/main/java/com/salesforce/mirus/KafkaMonitor.java [hunk 13]
```diff
@@ -89,6 +97,52 @@
...
+          regexRouters.add(transform);
+          // Slightly awkward check to see if any other routing transforms are configured
+        } else if (transformName.contains("Router")) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Unsupported Router Transformation %s found."
+                      + " To use it, please disable destination topic checking by setting 'enable.destination.topic.checking' to false.",
+                  transformName));
+        } else {
+          logger.debug("Ignoring non-routing Transform %s", transformName);
```
Comment (adammelliott) [inline src/main/java/com/salesforce/mirus/KafkaMonitor.java]: Logger interpolation placeholder is `{}`.

**Comment by parafiend** [comment]
@pdavidson100 Fixed minor typos, updated draft release notes, and reworked into more cohesive commits. Please approve for merging.
